libgoal/transactions.go

```diff
@@ -220,3 +220,64 @@ func (c *Client) MakeUnsignedGoOfflineTx(address string, round, txValidRounds, f
 	}
 	return goOfflineTransaction, nil
 }
+
+// FillUnsignedTxTemplate fills in header fields in a partially-filled-in transaction.
+func (c *Client) FillUnsignedTxTemplate(sender string, firstValid, numValidRounds, fee uint64, tx transactions.Transaction) (transactions.Transaction, error) {
+	// Parse the address
+	parsedAddr, err := basics.UnmarshalChecksumAddress(sender)
+	if err != nil {
+		return transactions.Transaction{}, err
+	}
+
+	params, err := c.SuggestedParams()
+	if err != nil {
+		return transactions.Transaction{}, err
+	}
+
+	cparams, ok := config.Consensus[protocol.ConsensusVersion(params.ConsensusVersion)]
+	if !ok {
+		return transactions.Transaction{}, errors.New("unknown consensus version")
+	}
+
+	// Determine the last round this tx will be valid
+	if firstValid == 0 {
+		firstValid = params.LastRound + 1
+	}
+
+	if numValidRounds == 0 {
+		numValidRounds = cparams.MaxTxnLife
+	}
+
+	parsedFirstValid := basics.Round(firstValid)
+	parsedLastValid := basics.Round(firstValid + numValidRounds)
+	parsedFee := basics.MicroAlgos{Raw: fee}
+
+	tx.Header = transactions.Header{
+		Sender:     parsedAddr,
+		Fee:        parsedFee,
+		FirstValid: parsedFirstValid,
+		LastValid:  parsedLastValid,
+	}
+
+	if cparams.SupportGenesisHash {
+		var genHash crypto.Digest
+		copy(genHash[:], params.GenesisHash)
+		tx.GenesisHash = genHash
+	}
+
+	// Default to the suggested fee, if the caller didn't supply it
+	// Fee is tricky, should taken care last. We encode the final
+	// transaction to get the size post signing and encoding.
+	// Then, we multiply it by the suggested fee per byte.
+	if fee == 0 {
+		tx.Fee = basics.MulAIntSaturate(basics.MicroAlgos{Raw: params.Fee}, tx.EstimateEncodedSize())
```

Comment:
how does msgpack encodes the fee ? I'm asking since if it's not a fixed byte-length, we might need to adjust the `tx.EstimateEncodedSize` according to the fee.

Reply:
The encoding is variable-length: it's zero extra bytes for a zero value, map key length + 1 byte for a value up to 127, map key length + 2 bytes for a value up to 255, map key length + 3 bytes for a value up to 65535, map key length + 5 bytes for up to 2^32-2, and otherwise map key length + 9 bytes.  There might also be an extra byte added if the entire transaction object grows from 15 to 16 non-zero-valued keys (though I think that's unlikely).

That's one reason why the method is called `EstimateEncodedSize`.